starlette/testclient.py

```diff
@@ -81,26 +79,22 @@ def _get_reason_phrase(status_code: int) -> str:
         return ""
 
 
-def _is_asgi3(app: typing.Union[ASGI2App, ASGI3App]) -> bool:
-    if inspect.isclass(app):
-        return hasattr(app, "__await__")
-    elif inspect.isfunction(app):
-        return asyncio.iscoroutinefunction(app)
-    call = getattr(app, "__call__", None)
-    return asyncio.iscoroutinefunction(call)
-
-
 class _WrapASGI2:
     """
     Provide an ASGI3 interface onto an ASGI2 app.
     """
 
-    def __init__(self, app: ASGI2App) -> None:
+    def __init__(self, app: typing.Union[ASGI2App, ASGI3App]) -> None:
         self.app = app
 
     async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
-        instance = self.app(scope)
-        await instance(receive, send)
+        try:
+            asgi3app = typing.cast(ASGI3App, self.app)
+            aw = asgi3app(scope, receive, send)
+        except TypeError:
+            asgi2app = typing.cast(ASGI2App, self.app)
+            aw = asgi2app(scope)(receive, send)
+        await aw
```

Comment:
Do we need to support ASGI 2.0 anymore?
Perhaps we could look at dropping that to keep things more obvious?

(I don't see any great resource in the ASGI docs on the 2 -> 3 changes?)

Reply:
here: https://asgi.readthedocs.io/en/latest/specs/main.html?highlight=2.0#legacy-applications

Reply:
Based on that it seems reasonable for TestClient to drop support for ASGI 2.0

Reply:
I think dropping ASGI 2 first would be pretty reasonable, yup.